File: neetcode150/graphs/leetcode994.py
from collections import deque
from typing import List


class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        """
        You are given an m x n grid where each cell can have one of three values:
            * 0 representing an empty cell,
            * 1 representing a fresh orange, or
            * 2 representing a rotten orange.

        Every minute, any fresh orange that is 4-directionally adjacent to a rotten orange becomes rotten.

        Return the minimum number of minutes that must elapse until no cell has a fresh orange.
        If this is impossible, return -1.

        :param grid:
        :return:
        """
        fresh = 0
        minutes = 0
        queue = deque()
        rows, cols = len(grid), len(grid[0])

        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1:
                    fresh += 1

                if grid[i][j] == 2:
                    queue.append((i, j))

        if fresh == 0:
            return 0

        while queue and fresh > 0:
            minutes += 1
            size = len(queue)

            for _ in range(size):
                i, j = queue.popleft()

                for di, dj in directions:
                    ci, cj = i + di, j + dj
                    if (
                        ci < 0
                        or cj < 0
                        or ci >= rows
                        or cj >= cols
                        or grid[ci][cj] != 1
                    ):
                        continue

                    fresh -= 1

                    grid[ci][cj] = 2

                    queue.append((ci, cj))

        return minutes if fresh == 0 else -1

File: js_algorithms/neetcode150/graphs/leetcode994.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
const orangesRotting = function (grid) {
  // count the number of fresh oranges
  // we can start making things rotten by doing a BFS around each rotten tomato
  const rows = grid.length;
  const cols = grid[0].length;
  let minutes = 0;
  let fresh = 0;
  let queue = [];

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] === 1) {
        fresh++;
      }

      if (grid[i][j] === 2) {
        queue.push([i, j]);
      }
    }
  }

  if (fresh === 0) {
    return 0;
  }

  while (queue.length) {
    const size = queue.length;
    minutes++;

    for (let i = 0; i < size; i++) {
      const [cx, cy] = queue.shift();

      for (
        const [dx, dy] of [
          [1, 0],
          [-1, 0],
          [0, 1],
          [0, -1],
        ]
      ) {
        const [nx, ny] = [cx + dx, cy + dy];

        if (
          nx < 0 ||
          ny < 0 ||
          nx >= rows ||
          ny >= cols ||
          grid[nx][ny] !== 1
        ) {
          continue;
        }

        grid[nx][ny] = 2;
        fresh--;

        queue.push([nx, ny]);
      }
    }
  }

  return fresh === 0 ? minutes - 1 : -1;
};
